core/src/main/java/com/scalar/db/storage/dynamo/QueryScanner.java

```diff
@@ -49,18 +58,23 @@ private boolean hasNext() {
       return true;
     }
     if (lastEvaluatedKey != null) {
-      handleResponse(request.execute(lastEvaluatedKey));
+      if (remainingLimit != null) {
+        handleResponse(request.execute(lastEvaluatedKey, remainingLimit));
+      } else {
+        handleResponse(request.execute(lastEvaluatedKey));
+      }
       return itemsIterator.hasNext();
     }
     return false;
   }
 
   private void handleResponse(PaginatedRequestResponse response) {
     List<Map<String, AttributeValue>> items = response.items();
-    totalResultCount += items.size();
+    if (remainingLimit != null) {
+      remainingLimit -= items.size();
+    }
     itemsIterator = items.iterator();
-    if ((request.limit() == null || totalResultCount < request.limit())
-        && response.hasLastEvaluatedKey()) {
+    if ((remainingLimit == null || remainingLimit > 0) && response.hasLastEvaluatedKey()) {
```

Comment:
I'm confused so I'm probably wrong, but should it be something like below?
```suggestion
    if ((remainingLimit != null && remainingLimit > 0) && response.hasLastEvaluatedKey()) {
```
I'm probably a bit confused with the meaning of null `remainingLimit`.

Reply:
The logic is as follows:

- If `remainingLimit` is `null` (meaning the user hasn’t specified a limit in Scan), and the response includes the last evaluated key, we need to send an additional request to DynamoDB.
- If `remainingLimit` is not `null` (meaning the user has specified a limit in Scan) and `remainingLimit` is greater than `0`, and if the response includes the last evaluated key, we also need to send an additional request to DynamoDB.
- Otherwise, no further requests to DynamoDB are necessary.

So I think the current code is correct.

Reply:
I got a better understanding. Thanks!